scripting/src/vss.ts

```diff
@@ -0,0 +1,609 @@
+export type VssQuantResult<K extends keyof VssQuantMap> = void | undefined | "preventDefault" | VssQuantMap[K][1];
+
+export interface VssQuantMap {
+    "ready": [void, void],
+    "runtime_object": [VssRuntimeObjectQuant, void],
+    "scaled_renderer": [VssScaledRendereQuant, void],
+    "option": [VssOptionQuant, VssOptionQuantResult],
+    "camera": [VssCameraQuant, VssCameraQuantResult],
+    "mechos_traction": [VssMechosTractionQuant, VssMechosTractionQuantResult],
+}
+
+export interface VssRuntimeObjectQuant {
+    runtimeObjectId: RoadRuntimeObjId,
+}
+
+export interface VssScaledRendereQuant {
+    enabled: boolean,
+}
+
+export interface VssOptionQuant {
+    id: iScreenOptionId,
+    value: number,
+}
+
+export interface VssOptionQuantResult {
+    value?: number;
+}
+
+export interface VssCameraQuant {
+    turnAngle: number,
+    slopAngle: number,
+}
+export interface VssCameraQuantResult {
+    turnAngle?: number;
+    slopeAngle?: number;
+}
+
+export interface VssMechosTractionQuant {
+    traction: number,
+    rudder: number,
+    tractionIncrement: number,
+    tractionDecrement: number,
+    tractionMax: number,
+    rudderStep: number,
+    rudderMax: number,
+    unitAngle: number,
+    helicopterStrife: number,
+}
+export interface VssMechosTractionQuantResult {
+    traction?: number,
+    rudder?: number,
+    helicopterStrife?: number,
+}
+
+export type VssQuantListener<K extends VssQuantName> = (payload: VssQuantPayload<K>,
+    stopPropogation: () => void, quant: K) => VssQuantResult<K>;
+
+export class VssMath {
+    PI = 1 << 11;
+    PI_2 = this.PI / 2;
+
+    angleToRadians(angle: number) {
+        return angle * Math.PI / this.PI;
+    }
+
+    radiansToAngle(radians: number) {
+        return Math.round(radians * this.PI / Math.PI);
+    }
+}
+
+class Vss {
+    // all scripts that available to use in 'require'
+    scripts: string[];
+
+    // game math
+    math = new VssMath();
+
+    private quantListeners: { [quantName: string]: VssQuantListener<any>[] } = {};
+
+    constructor() {
+        this.scripts = bridge.scripts().filter((value) => {
+            return value !== "vss.js" && value !== "main.js";
+        }).map((value) => value.substring(0, value.length - 3));
+
+        global.onVssQuant = this.onVssQuant.bind(this);
+    }
+
+    fatal(msg: string) {
+        bridge.fatal("vss: " + msg);
+    }
+
+    isKeyPressed(scanCode: number) {
+        return bridge.isKeyPressed(scanCode);
+    }
+
+    addQuantListener<K extends VssQuantName>(quant: K, listener: VssQuantListener<K>) {
+        if (this.quantListeners[quant] === undefined) {
+            this.quantListeners[quant] = [];
+        }
+        this.quantListeners[quant].push(listener);
+    }
+
+    removeQuantListener<K extends VssQuantName>(quant: K, listener: VssQuantListener<K>) {
+        const index = this.quantListeners[quant]?.indexOf(listener);
+        if (index && index !== -1) {
```

Comment:
I believe you want to check `index` for equals to `undefined` strictly ( `index === void 0 && index !== -1` ).

Reply:
:+1: 
more readable will be `index !== undefined`